### PR TITLE
Standard Run Fixes

### PR DESCRIPTION
**Pull request by jcaravaca**
Improvements and bug fixes for the Standard Runs and ECA panels.
Send notifications to main thread.
Do not store detector state bits in run type word for the Standard Runs.
Delete estop buttons in hvMaster.

### CODE REVIEW

**Comment by jcaravaca** [comment]
@tlatorre-uchicago this should be ready to go now

File: Source/Experiments/SNOP/SNOPModel.m [hunk 1]
```diff
@@ -1457,8 +1395,8 @@ - (void)encodeWithCoder:(NSCoder*)encoder
     [encoder encodeInt:[self logPort] forKey:@"logPort"];
 
     //Runs tab
-    [encoder encodeObject:[self standardRunType] forKey:@"SNOPStandardRunType"];
-    [encoder encodeObject:[self standardRunVersion] forKey:@"SNOPStandardRunVersion"];
+    //[encoder encodeObject:[self standardRunType] forKey:@"SNOPStandardRunType"];
+    //[encoder encodeObject:[self standardRunVersion] forKey:@"SNOPStandardRunVersion"];
```
Comment (tlatorre-uchicago) [inline Source/Experiments/SNOP/SNOPModel.m]: Why are these lines commented out? Is the plan not to save the standard run settings to the Orca file?

File: Source/Experiments/SNOP/SNOPModel.m [hunk 2]
```diff
@@ -185,8 +183,10 @@ - (void) setMTCPort: (int) port
         [caen setMTCPort:port];
     }
 
-    [[NSNotificationCenter defaultCenter]
-     postNotificationName:@"SNOPSettingsChanged" object:self];
+    dispatch_async(dispatch_get_main_queue(), ^{
```
Comment (tlatorre-uchicago) [inline Source/Experiments/SNOP/SNOPModel.m]: Why did you change all of the setters to run on the main thread?

File: Source/Experiments/SNOP/SNOPController.m [hunk 3]
```diff
@@ -1630,12 +1638,23 @@ -(void) displayThresholdsFromDB {
...
-        NSLogColor([NSColor redColor],@"Standard Run Version not set \n");
+        NSLogColor([NSColor redColor],@"Standard Run not set (ignore this if you just restarted ORCA) \n");
+        return;
+    }
+    //If no version: display null values and quit
+    if([model standardRunVersion] == nil || [[model standardRunVersion] isEqualToString:@""]){
+        for (int i=0; i<[standardRunThresStoredValues numberOfRows];i++) {
+            [[standardRunThresStoredValues cellAtRow:i column:0] setStringValue:@"--"];
+        }
+        NSLogColor([NSColor redColor],@"Test Run not set (ignore this if you just restarted ORCA) \n");
```
Comment (tlatorre-uchicago) [inline Source/Experiments/SNOP/SNOPController.m]: I think showing red log messages on every start up isn't a good idea since it will lead to alarm fatigue, i.e. operators will get used to red alerts whenever starting ORCA. What about just not loading or displaying thresholds until the user clicks the load values from DB button?

**Comment by tlatorre-uchicago** [comment]
@jcaravaca I've added a few comments. After addressing these, I'll test it out on the teststand.

File: Source/Experiments/SNOP/SNOPModel.m (hunk 1 above)
Comment (jcaravaca) [inline Source/Experiments/SNOP/SNOPModel.m]: That'd be neat, but then it conflicts with the fact that we want to reload the SR only when clicking the reload button. To be more specific: now when you restart ORCA, it doesn't contact the SR RB automatically, but you need to press 'reload'. In this case, it doesn't make sense to save those fields in the .ORCA file.

File: Source/Experiments/SNOP/SNOPModel.m (hunk 2 above)
Comment (jcaravaca) [inline Source/Experiments/SNOP/SNOPModel.m]: Some setters send notifications to change the GUI and it is recommended to send them to the main thread. That might be what was causing the SNOP GUI to freeze. The XL3 object does the same thing.

File: Source/Experiments/SNOP/SNOPModel.m (hunk 1 above)
Comment (tlatorre-uchicago) [inline Source/Experiments/SNOP/SNOPModel.m]: Yeah, it looks like a lot of methods in the XL3 Link controller do that, but I think we should instead force code running in separate threads to call these methods on the main function instead of putting it in the setter itself, otherwise we are going to incur the overhead of putting these on the queue for every setter method and the fact that it is called as dispatch_async() means that the variable actually isn't guaranteed to be set when you reach the next line after calling the setter.

File: Source/Experiments/SNOP/SNOPModel.m (hunk 2 above)
Comment (tlatorre-uchicago) [inline Source/Experiments/SNOP/SNOPModel.m]: Yeah, it looks like a lot of methods in the XL3 Link controller do that, but I think we should instead force code running in separate threads to call these methods on the main function instead of putting it in the setter itself, otherwise we are going to incur the overhead of putting these on the queue for every setter method and the fact that it is called as dispatch_async() means that the variable actually isn't guaranteed to be set when you reach the next line after calling the setter.

File: Source/Experiments/SNOP/SNOPController.m (hunk 3 above)
Comment (jcaravaca) [inline Source/Experiments/SNOP/SNOPController.m]: This is a bit tricky: this function is called by SRVersionChanged, which is called when that SR field has changed OR the SNOP window is opened. If no SR field name is selected this warning pops up and the display will show dashes rather than the numbers from the DB.
This needs to be like this because not only the user can change the SR field (e.g. RunScripts do as well), so I think the logic needs to be kept like this.
On the other hand, I could be more explicit on the message and show it in black color, as technically it might not be a warning:
'Standard Run not set: reload the DB and select a SR in the dropdown menu'

Comment (tlatorre-uchicago) [inline Source/Experiments/SNOP/SNOPController.m]: In that case, it's not an error at all right? I think in that case, no log message is needed. Can you elaborate in the comments the difference between nil and the empty string? It's not clear what the two mean.

**Comment by jcaravaca** [comment]
@tlatorre-uchicago I've implemented the minimal changes we were talking about last time. I think it's worth to merge this if there are no major issues to unlock the GUI development.

File: Source/Experiments/SNOP/SNOPModel.m (hunk 2 above)
Comment (tlatorre-uchicago) [inline Source/Experiments/SNOP/SNOPModel.m]: This dispatch_async still needs to be removed.

File: Source/Experiments/SNOP/SNOPModel.m [hunk 4]
```diff
@@ -225,8 +225,10 @@ - (void) setMTCHost: (NSString *) host
         [caen setMTCHost:host];
     }
 
-    [[NSNotificationCenter defaultCenter]
-     postNotificationName:@"SNOPSettingsChanged" object:self];
+    dispatch_async(dispatch_get_main_queue(), ^{
```
Comment (tlatorre-uchicago) [inline Source/Experiments/SNOP/SNOPModel.m]: here too.

File: Source/Experiments/SNOP/SNOPModel.m [hunk 5]
```diff
@@ -243,8 +245,9 @@ - (void) setXL3Port: (int) port
     [xl3_server disconnect];
     [xl3_server setPort:port];
 
-    [[NSNotificationCenter defaultCenter]
-     postNotificationName:@"SNOPSettingsChanged" object:self];
+    dispatch_async(dispatch_get_main_queue(), ^{
```
Comment (tlatorre-uchicago) [inline Source/Experiments/SNOP/SNOPModel.m]: here too.

File: Source/Experiments/SNOP/SNOPModel.m [hunk 6]
```diff
@@ -275,8 +278,10 @@ - (void) setXL3Host: (NSString *) host
         [[xl3 xl3Link] disconnectSocket];
     }
 
-    [[NSNotificationCenter defaultCenter]
-     postNotificationName:@"SNOPSettingsChanged" object:self];
+    dispatch_async(dispatch_get_main_queue(), ^{
```
Comment (tlatorre-uchicago) [inline Source/Experiments/SNOP/SNOPModel.m]: here

File: Source/Experiments/SNOP/SNOPModel.m [hunk 7]
```diff
@@ -648,7 +650,10 @@ - (void) runInitialization:(NSNotification*)aNote
                                 @"", @"Reason",
                                 @"", @"Details",
                                 nil];
-    [[NSNotificationCenter defaultCenter] postNotificationName:ORAddRunStartupAbort object: self userInfo: userInfo];
+    dispatch_async(dispatch_get_main_queue(), ^{
```
Comment (tlatorre-uchicago) [inline Source/Experiments/SNOP/SNOPModel.m]: here

File: Source/Experiments/SNOP/SNOPModel.m [hunk 8]
```diff
@@ -703,7 +708,9 @@ - (void) runAboutToStart:(NSNotification*)aNote
...
```
Comment (tlatorre-uchicago) [inline Source/Experiments/SNOP/SNOPModel.m]: here

File: Source/Experiments/SNOP/SNOPModel.m [hunk 9]
```diff
@@ -752,7 +759,9 @@ - (void) runStarted:(NSNotification*)aNote
                                 @"", @"Reason",
                                 @"", @"Details",
                                 nil];
-    [[NSNotificationCenter defaultCenter] postNotificationName:ORRequestRunHalt object: self userInfo: userInfo];
+    dispatch_async(dispatch_get_main_queue(), ^{
```
Comment (tlatorre-uchicago) [inline Source/Experiments/SNOP/SNOPModel.m]: here

File: Source/Experiments/SNOP/SNOPModel.m [hunk 10]
```diff
@@ -784,8 +793,9 @@ - (void) runAboutToStop:(NSNotification*)aNote
         NSDictionary *userInfo = [NSDictionary dictionaryWithObjectsAndKeys:
                                   @"waiting for MTC/XL3/CAEN data", @"Reason",
                                   nil];
-
-        [[NSNotificationCenter defaultCenter] postNotificationName:ORAddRunStateChangeWait object: self userInfo: userInfo];
+        dispatch_async(dispatch_get_main_queue(), ^{
```
Comment (tlatorre-uchicago) [inline Source/Experiments/SNOP/SNOPModel.m]: here

File: Source/Experiments/SNOP/SNOPModel.m [hunk 11]
```diff
@@ -1082,7 +1094,10 @@ - (void) setOrcaDBIPAddress:(NSString*)orcaIPAddress
 		[[NSUserDefaults standardUserDefaults] setObject:self.orcaDBConnectionHistory forKey:[NSString stringWithFormat:@"orca.%@.orcaDBConnectionHistory",[self className]]];
 		[[NSUserDefaults standardUserDefaults] setInteger:self.orcaDBIPNumberIndex forKey:[NSString stringWithFormat:@"orca.%@.orcaDBIPNumberIndex",[self className]]];
 		[[NSUserDefaults standardUserDefaults] synchronize];
-		[[NSNotificationCenter defaultCenter] postNotificationName:ORSNOPModelOrcaDBIPAddressChanged object:self];
+        dispatch_async(dispatch_get_main_queue(), ^{
```
Comment (tlatorre-uchicago) [inline Source/Experiments/SNOP/SNOPModel.m]: Not sure where this is called from, but I think running this on the main thread should be done separately if anyone is trying to change it outside of the main thread, and not in the setter.

File: Source/Experiments/SNOP/SNOPModel.m [hunk 12]
```diff
@@ -1113,7 +1128,9 @@ - (void) setDebugDBIPAddress:(NSString*)debugIPAddress
 		[[NSUserDefaults standardUserDefaults] setObject:self.debugDBConnectionHistory forKey:[NSString stringWithFormat:@"orca.%@.debugDBConnectionHistory",[self className]]];
 		[[NSUserDefaults standardUserDefaults] setInteger:self.debugDBIPNumberIndex forKey:[NSString stringWithFormat:@"orca.%@.debugDBIPNumberIndex",[self className]]];
 		[[NSUserDefaults standardUserDefaults] synchronize];
-		[[NSNotificationCenter defaultCenter] postNotificationName:ORSNOPModelDebugDBIPAddressChanged object:self];
+        dispatch_async(dispatch_get_main_queue(), ^{
```
Comment (tlatorre-uchicago) [inline Source/Experiments/SNOP/SNOPModel.m]: same here

**Comment by tlatorre-uchicago** [comment]
@jcaravaca I've added comments for lines where you've added asynchronous dispatches to the main thread that I think need to be removed. If you think I missed something and these are correct changes, let me know. Otherwise, after they are removed I'll test on site and merge.

**Comment by jcaravaca** [comment]
I think the 'dispatchs' came back as an artifact of the rebase with master. Now they should be gone except the one in '_waitForBuffers'.

**Comment by tlatorre-uchicago** [comment]
merged.
